File: 期末/上机/02_MATLAB算法模板/02_Runge_Kutta/T03_Explicit_Runge_Kutta.m
%% 一般显式 Runge-Kutta：用 Butcher 表覆盖三阶 Heun、三阶 RK、经典 RK4
clear; clc; close all;                                      % 清理环境。
f = @(t,y) -y+t.^2+1;                                      % 【必须替换】标量或列向量 ODE 右端。
t0 = 0; T = 1; y0 = 1; h = 0.1;                           % 【必须替换】区间、初值、步长；向量初值写成 [a;b]。
method = "RK4";                                            % 【必须替换】可选 Heun3/RK3/RK4。
switch method                                              % 根据方法装入 Butcher 系数 A,b,c。
    case "Heun3"                                           % 教材三级三阶 Heun 方法。
        A = [0 0 0; 1/3 0 0; 0 2/3 0];                    % A 严格下三角，所以阶段可依次显式计算。
        b = [1/4;0;3/4]; c = [0;1/3;2/3];                 % b 是最终权重，c 是阶段时间位置。
    case "RK3"                                             % 教材公式 (5.3.13) 的三级三阶 RK。
        A = [0 0 0; 1/2 0 0; 0 3/4 0];                    % 每行给出该阶段对前面阶段的组合。
        b = [2/9;3/9;4/9]; c = [0;1/2;3/4];               % 权重和为 1，保证至少一阶相容。
    case "RK4"                                             % 最常用的经典四级四阶 RK。
        A = [0 0 0 0; 1/2 0 0 0; 0 1/2 0 0; 0 0 1 0];    % 经典 RK4 的阶段矩阵。
        b = [1/6;1/3;1/3;1/6]; c = [0;1/2;1/2;1];         % 经典 RK4 权重与节点。
    otherwise                                              % 未知名称时停止。
        error('未知 method：%s',method);                    % 显示错误方法名。
end                                                        % Butcher 表选择结束。
N = round((T-t0)/h);                                       % 计算总步数。
assert(abs(t0+N*h-T)<1e-12,'步长必须整除区间长度。');       % 检查终点。
t = t0+(0:N)*h;                                            % 生成等距时间网格。
y0 = y0(:);                                                % 冒号把标量/行向量统一成列向量。
y = zeros(numel(y0),N+1); y(:,1) = y0;                    % 每一列保存一个时间点的状态向量。
s = numel(b);                                              % RK 级数等于权重个数。
for n = 1:N                                                % 逐步推进。
    K = zeros(numel(y0),s);                                % K(:,r) 保存第 r 个阶段斜率。
    for r = 1:s                                            % 依次计算所有显式阶段。
        stageY = y(:,n)+h*K*A(r,:).';                      % 矩阵乘法组合已算阶段；.' 是非共轭转置。
        K(:,r) = f(t(n)+c(r)*h,stageY);                    % 在阶段点计算斜率 f。
    end                                                    % 阶段计算结束。
    y(:,n+1) = y(:,n)+h*K*b;                               % 用权重 b 加权所有阶段，完成一步。
end                                                        % 时间推进结束。
disp(table(t.',y(1,:).','VariableNames',{'t','第一分量'})); % 展示第一分量；方程组可自行增加其他列。
plot(t,y(1,:),'o-','LineWidth',1.3); grid on; xlabel('t'); ylabel('y_1'); title(method); % 画第一分量。

File: 期末/上机/03_Python算法模板/02_Runge_Kutta/T03_Explicit_Runge_Kutta.py
# 一般显式 Runge-Kutta：用 Butcher 表覆盖三阶 Heun、三阶 RK 和经典 RK4。  # 数值分析：显式 RK 的 A 矩阵严格下三角。
import numpy as np  # NumPy 用于向量状态、矩阵乘法和时间网格。
import matplotlib.pyplot as plt  # Matplotlib 用于画第一分量。

f = lambda t, y: -y + t**2 + 1  # 【必须替换】标量或向量 ODE 右端；向量题应返回 NumPy 一维数组。
t0, T, y0, h = 0.0, 1.0, 1.0, 0.1  # 【必须替换】区间、初值、步长；向量初值写 np.array([a,b])。
method = "RK4"  # 【必须替换】可选 Heun3/RK3/RK4。
if method == "Heun3":  # 教材三级三阶 Heun 方法。
    A = np.array([[0,0,0],[1/3,0,0],[0,2/3,0]], dtype=float)  # A 的第 r 行表示第 r 阶段对前面阶段的组合。
    b = np.array([1/4,0,3/4], dtype=float)  # b 是最后一步组合各阶段的权重。
    c = np.array([0,1/3,2/3], dtype=float)  # c 是各阶段所在的相对时间位置。
elif method == "RK3":  # 教材公式 (5.3.13) 的三级三阶 RK。
    A = np.array([[0,0,0],[1/2,0,0],[0,3/4,0]], dtype=float)  # 严格下三角结构保证阶段可依次算出。
    b = np.array([2/9,3/9,4/9], dtype=float)  # 最终权重和为 1，保证相容性。
    c = np.array([0,1/2,3/4], dtype=float)  # 阶段节点。
elif method == "RK4":  # 经典四级四阶 Runge-Kutta。
    A = np.array([[0,0,0,0],[1/2,0,0,0],[0,1/2,0,0],[0,0,1,0]], dtype=float)  # 经典 RK4 阶段矩阵。
    b = np.array([1/6,1/3,1/3,1/6], dtype=float)  # 经典 RK4 最终权重。
    c = np.array([0,1/2,1/2,1], dtype=float)  # 经典 RK4 阶段时间位置。
else:  # 捕获未知方法名。
    raise ValueError(f"未知 method：{method}")  # 抛出异常并显示错误方法名。
N = round((T-t0)/h)  # 计算总步数。
assert abs(t0+N*h-T) < 1e-12, "步长必须整除区间长度。"  # 检查固定网格终点。
t = t0 + np.arange(N+1)*h  # 创建等距时间网格。
y0 = np.atleast_1d(np.asarray(y0, dtype=float))  # NumPy 语法：把标量或列表统一成一维浮点数组。
y = np.zeros((N+1, y0.size), dtype=float)  # 每一行保存一个时间点的状态向量。
y[0] = y0  # 写入初始状态。
s = b.size  # RK 级数等于权重向量长度。
for n in range(N):  # 逐步推进。
    K = np.zeros((s, y0.size), dtype=float)  # K[r] 保存第 r 个阶段的斜率向量。
    for r in range(s):  # 依次计算显式阶段。
        stage_y = y[n] + h*(A[r] @ K)  # Python/NumPy 语法：@ 表示矩阵乘法；组合已算阶段。
        K[r] = np.atleast_1d(f(t[n]+c[r]*h, stage_y))  # 在阶段点计算斜率，并统一为一维数组。
    y[n+1] = y[n] + h*(b @ K)  # 按权重 b 组合全部阶段，完成一步。
for tn, yn in zip(t, y[:,0]):  # 只打印第一分量；方程组可自行增加其他分量。
    print(f"t={tn:.3f}, y1={yn:.10f}")  # 格式化输出第一分量。
plt.plot(t, y[:,0], "o-", linewidth=1.3, label=method)  # 画第一分量数值解。
plt.xlabel("t")  # 设置横轴名称。
plt.ylabel("y[0]")  # 设置纵轴名称。
plt.grid(True)  # 打开网格。
plt.legend()  # 显示图例。
plt.tight_layout()  # 自动调整布局。
plt.show()  # 显示图像。
